---
title: Errors
slug: en/v2/guides/errors
---

## Introduction

Some error objects may have a `.cause` property. In most cases, this is of type `unknown`, and the value may change without notice in the future.

## General

### SurrealError

`SurrealError` is the base class that most error objects explicitly thrown by this SDK inherit from. This class inherits from JavaScript's `Error` class. This error object is rarely thrown alone, and usually a special error object with its own meaning is thrown by the inheriting class. It is recommended to use this error object to easily classify errors caught using `try-catch` syntax as errors explicitly thrown by this SDK.

#### Properties

`.name: "SurrealError"`
:   The name of the error.

`.message`
:   The error message.

`.stack`
:   Stack trace may be recorded.

`.cause`
:   The cause or context in which the error occurred may be set.

### SurrealTypeError

Inherits from: `SurrealError`

This is an error thrown when input values or other validations fail.

#### Properties

`.name: "SurrealTypeError"`
:   The name of the error.

`.expected`
:   The expected data format.

`.actual`
:   The string representation of the actual value.

#### Resolution

Trace the stack trace to find the reason why the expected data format is not being input.

### SurrealValueError

Inherits from: `SurrealError`

This is an error thrown when input values or other validations fail.

#### Properties

`.name: "SurrealValueError"`
:   The name of the error.

`.expected`
:   The expected data format.

`.actual`
:   The actual value.

#### Resolution

Trace the stack trace to find the reason why the expected data format is not being input.

### SurrealAggregateError

Inherits from: `SurrealError`

This is an error that aggregates multiple errors or error messages.

#### Properties

`.name: "SurrealAggregateError"`
:   The name of the error.

`.cause: unknown[]`
:   Each element of the array contains an error or an error message.

### CircularReferenceError

Inherits from: `SurrealError`

This is an error thrown when a circular reference is detected. It is mainly thrown when converting a JavaScript value to another format, such as when a parent object is detected to be the same object.

#### Properties

`.name: "CircularReferenceError"`
:   The name of the error.

`.reference: unknown`
:   The value where the circular reference occurred.

For example, this error is thrown in the following case:

```ts
import { toSurql } from "@tai-kun/surrealdb/utils";

const a = {};
a.a = a;

console.log(a); // <ref *1> { a: [Circular *1] }

toSurql(a); // throws CircularReferenceError
```

#### Resolution

This error can be thrown even in cases other than the `toSurql` function. You need to carefully debug whether the object contains the object itself.

### NumberRangeError

Inherits from: `SurrealError`

This is an error thrown when a number is outside the expected range.

#### Properties

`.name: "NumberRangeError"`
:   The name of the error.

`.range`
:   The expected range.

`.actual`
:   The actual value.

`.integer`
:   `true` if an integer is expected.

#### Resolution

Trace the stack trace to find the reason why the expected data format is not being input.

### UnsupportedRuntimeError

Inherits from: `SurrealError`

This is an error thrown when an unsupported runtime is detected.

#### Properties

`.name: "UnsupportedRuntimeError"`
:   The name of the error.

#### Resolution

You need to avoid using that runtime or polyfill it and test it thoroughly.

### UnreachableError

Inherits from: `SurrealError`

This is an error thrown when unreachable code is reached. If this error is thrown, you have probably encountered a bug in this SDK.

#### Properties

`.name: "UnreachableError"`
:   The name of the error.

#### Resolution

You can raise this issue from the following URL:

[https://github.com/tai-kun/surrealdb.js/issues](https://github.com/tai-kun/surrealdb.js/issues)

## Client

### CircularEngineReferenceError

Inherits from: `CircularReferenceError`

This is an error thrown when a circular reference occurs between engines when connecting to the database.

#### Properties

`.name: "CircularEngineReferenceError"`
:   The name of the error.

`.reference: string`
:   The scheme name where the circular reference occurred.

`.seen: string[]`
:   The list of scheme names that were traversed before the circular reference occurred.

For example, this error is thrown in the following case:

```ts
import { initSurreal } from "@tai-kun/surrealdb";

const { Surreal } = initSurreal({
  engines: {
    http: "https",
    https: "http",
  },
  // ...
});

await using db = new Surreal();
await db.connect("https://localhost:8000"); // throws CircularEngineReferenceError: Circular engine reference: http,https
```

In the example above, the https protocol engine is trying to use the engine that is set to http, but the http protocol engine is trying to use the engine that is set to https, so a circular reference error is thrown.

#### Resolution

Instead of setting the string-specified protocol name to the value of another protocol name, replace it with a concrete implementation that creates an engine:

```ts del={6} ins={7-10}
import { initSurreal } from "@tai-kun/surrealdb";
import HttpEngine from "@tai-kun/surrealdb/http-engine";

const { Surreal } = initSurreal({
  engines: {
    http: "https",
    http: config => new HttpEngine({
      ...config,
      // fetch: <your custom fetch function>
    }),
    https: "http",
  },
  // ...
});
```

### EngineNotFoundError

Inherits from: `SurrealError`

This is an error thrown when attempting to connect with a protocol that has no engine configured.

#### Properties

`.name: "EngineNotFoundError"`
:   The name of the error.

`.scheme: string`
:   The protocol name of the connection destination.

For example, this error is thrown in the following case:

```ts
import { initSurreal } from "@tai-kun/surrealdb";

const { Surreal } = initSurreal({
  engines: {
    http: config => new HttpEngine(config),
  },
  // ...
});

await using db = new Surreal();
await db.connect("https://localhost:8000"); // throws EngineNotFoundError: No https scheme engine found.
```

#### Resolution

Check the `engines` property to see if an engine can be created with the protocol of the connection destination endpoint.

```ts ins={6}
import { initSurreal } from "@tai-kun/surrealdb";

const { Surreal } = initSurreal({
  engines: {
    http: config => new HttpEngine(config),
    https: "http",
  },
  // ...
});
```

### ConnectionConflictError

Inherits from: `SurrealError`

This is an error thrown when the client attempts to connect to multiple endpoints simultaneously.

#### Properties

`.name: "ConnectionConflictError"`
:   The name of the error.

`.endpoint1: string`
:   The endpoint that is already connected.

`.endpoint2: string`
:   The other endpoint.

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

await db.connect("ws://localhost:8000");
await db.connect("ws://localhost:1129"); // throws ConnectionConflictError: An attempt was made to connect to ws://localhost:1129/rpc while ws://localhost:8000/rpc was already connected.
```

The `.connect` method appends `/rpc` to the end of the passed URL if the path does not end with `/rpc`. Therefore, if the URL path of one endpoint ends with `/rpc`, it may not throw an error despite its appearance:

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.connect("http://localhost:8000/rpc"); // OK!
```

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

await db.connect("http://localhost:8000/rpc");
await db.connect("http://localhost:8000"); // OK!
```

#### Resolution

You can connect to any endpoint by calling the `.close` method before calling the `.connect` method.

```ts ins={6}
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.close();
await db.connect("http://localhost:11298");
```

### MissingNamespaceError

Inherits from: `SurrealError`

This is an error thrown when a namespace is not selected before selecting a database. Or it is thrown even if you try to leave the namespace unselected after selecting a database.

#### Properties

`.name: "MissingNamespaceError"`
:   The name of the error.

`.database: string`
:   The database name.

For example, this error is thrown in the following case:

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.use({
  database: "example",
}); // throws MissingNamespaceError: The namespace must be specified before the database.
```

#### Resolution

Select the namespace as well when selecting a database.

```ts ins={7}
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.use({
  namespace: "example",
  database: "example",
});
```

### RpcResponseError

Inherits from: `SurrealError`

This is an error thrown when an RPC response indicates an error. This does not mean that there is a problem with communication or decoding of the response body due to the protocol connected, but it means that SurrealDB cannot process the RPC request.

#### Properties

`.name: "RpcResponseError"`
:   The name of the error.

`.id?: string`
:   The ID that identifies the RPC request. The ID always starts with `<method name>_`.

`.code: number`
:   Not specified in the SurrealDB documentation, but it is probably the JSON-RPC error code.

#### Resolution

The cause can vary, but it is possible that the version of SurrelDB supported by this SDK and the version of SurrelDB that processes the RPC request are different.

### QueryFailedError

Inherits from: `SurrealAggregateError`

This is an error thrown when the query fails.

#### Properties

`.name: "QueryFailedError"`
:   The name of the error.

`.cause: string[]`
:   List of errors.

For example, this error is thrown in the following case:

```ts
await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.query("OUTPUT 'Hello'"); // throws QueryFailedError: Query failed with 1 error(s)
```

#### Resolution

Make sure the SurrelQL syntax is correct.

### Closed

Inherits from: `SurrealError`

This is an error that indicates that the connection has been forcibly closed.

#### Properties

`.name: "Closed"`
:   The name of the error.

For example, you will get this error in the following case:

```ts
const db = new Surreal();

db.on("<event_name>", ({ signal }) => {
  return new Promise((resolve, reject) => {
    signal.addEventListener("abort", function() {
      console.error(this.reason); // Closed
      reject();
    })

    // ...
  });
});

await db.close({ force: true });
```

## Engine

### EngineError

Inherits from: `SurrealError`

This indicates that an error is due to the engine. It is dispatched via an event listener.

#### Properties

`.name: "EngineError"`
:   The name of the error.

`.fatal: boolean | undefined`
:   Indicates whether this error is fatal.

### HttpEngineError

Inherits from: `EngineError`

This indicates that an error is due to the HTTP engine. Currently, it is only defined and not used.

#### Properties

`.name: "HttpEngineError"`
:   The name of the error.

### WebSocketEngineError

Inherits from: `EngineError`

This indicates that an error is due to the WebSocket engine. It is used when parsing the response body fails before finding the ID that identifies the RPC request, when receiving an error event from the WebSocket object, when disconnecting the WebSocket, etc.

This SDK uses `315x` as a custom status code. This range is unassigned in the [IANA](https://www.iana.org/assignments/websocket/websocket.xml#close-code-number) status code.

#### Properties

`.name: "WebSocketEngineError"`
:   The name of the error.

`.code: number`
:   One of the following status codes:
    - `1000`~`1015`: [MDN Reference](https://developer.mozilla.org/docs/Web/API/CloseEvent/code)
    - `3150`: Indicates that the "error" event was received from the `WebSocket`[^1] instance.
    - `3151`: Indicates that an error occurred within the "open" event handler registered by the WebSocket engine. In most cases, `StateTransitionError` occurs during the transition from the connecting state to the connected state.
    - `3152`: Indicates that an error occurred within the "message" event handler registered by the WebSocket engine. There are various possible causes, but it is most likely that `ServerResponseError` or `RpcResponseError` was thrown within the event handler. It is possible that the request was made with an invalid RPC parameter. In that case, the method that called the RPC will (by default) time out after 5 seconds and fail. In versions prior to SurrealDB v2.0.2, timeouts may occur due to a wider range of causes.
    - `3153`: Indicates that the ping failed to be sent or received. This may be a temporary error, but if you are continuously receiving it, you may not be able to maintain the connection.
    - `3154`: Indicates that an error occurred within the "close" event handler registered by the WebSocket engine. In most cases, `StateTransitionError` occurs during the transition from the disconnecting state to the disconnected state.
:   Note that the following status codes are excluded from the above. This SDK does not treat them as errors.
    - `1000`: Closed normally.
    - `1001`: Early disconnects are common.
    - `1004`: Reserved.
    - `1005`: Reserved.
    - `1006`: Reserved.
    - `1015`: Reserved.

^1: `WebSocket` may be a class defined in the global variable by the runtime, or a class of `ws`.

#### Resolution

`1002`,`1003`,`1007`...`1011`,`1014`,`3150`,`3151`,`3154`
:   There is probably no way to automatically recover from these status codes. You need to fix the runtime, SurrealDB settings, or hard-coded implementation.

`3152`
:   Check if the arguments passed to the method (i.e. the contents of the RPC request) are correct.

`1012`,`1013`,`3153`
:   You may be able to recover automatically using the experimental feature [`autoReconnect()`](/v2/experimental/auto-reconnect).

### StateTransitionError

Inherits from: `SurrealAggregateError`

This is an error thrown when the execution of the event listener fails during state transition.

#### Properties

`.name: "StateTransitionError"`
:   The name of the error.

`.from: string`
:   The state at the start of the state transition.

`.to: string`
:   The destination state of the transition.

`.fallback: string`
:   The fallback transition destination if the state transition fails. If the value is the same as `.to`, it means that the transition was forced.

For example, this error is thrown in the following case:

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

db.on("connecting", () => {
  throw new Error("Don't let me connect !!!")
});

await db.connect("http://localhost:8000"); // throws ...
// StateTransitionError: The transition from `3` to `0` failed, falling back to `3`.
//     at <stack trace ...> {
//   cause: [
//     Error: Don't let me connect !!!
//         at <stack trace ...>
//   ],
//   from: 3,
//   to: 0,
//   fallback: 3
// }
```

### ConnectionUnavailableError

Inherits from: `SurrealError`

This is an error thrown when attempting to send an RPC request in an unconnected state.

#### Properties

`.name: "ConnectionUnavailableError"`
:   The name of the error.

### ServerResponseError

Inherits from: `SurrealError`

This is an error thrown when the response cannot be parsed as a PRC response. This is different from `RpcResponseError`.

#### Properties

`.name: "ServerResponseError"`
:   The name of the error.

#### Resolution

Usually, this error should not be thrown, but if it is thrown, the version of SurrelDB supported by this SDK and the version of SurrelDB that processes the RPC request may be different.

## CBOR

### CborError

Inherits from: `SurrealError`

This is the class inherited by all CBOR-related errors explicitly thrown by `@tai-kun/surrealdb/cbor`. This will not be thrown directly.

#### Properties

`.name: "CborError"`
:   The name of the error.

### CborWellFormednessError

Inherits from: `CborError`

#### Properties

This is the class inherited by all errors related to CBOR decoding explicitly thrown by `@tai-kun/surrealdb/cbor`. This will not be thrown directly.

See "Well-Formedness Errors and Examples" in Appendix F of RFC8949 [https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-](https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-)

`.name: "CborWellFormednessError"`
:   The name of the error.

### CborTooMuchDataError

Inherits from: `CborWellFormednessError`

This indicates that there are unconsumed input bytes remaining.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.2)

#### Properties

`.name: "CborTooMuchDataError"`
:   The name of the error.

#### Resolution

Check if the byte array to be decoded is in the correct CBOR format.

### CborTooLittleDataError

Inherits from: `CborWellFormednessError`

This indicates that the input bytes are incomplete.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.4)

#### Properties

`.name: "CborTooLittleDataError"`
:   The name of the error.

#### Resolution

Check if the byte array to be decoded is in the correct CBOR format.

### CborSyntaxError

Inherits from: `CborWellFormednessError`

This indicates that the input bytes do not match the CBOR encoding format.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.6)

#### Properties

`.name: "CborSyntaxError"`
:   The name of the error.

#### Resolution

Check if the byte array to be decoded is in the correct CBOR format.

### CborMaxDepthReachedError

Inherits from: `CborError`

This is an error thrown when the maximum depth of a JavaScript object is reached during CBOR encoding or decoding. The depth increases by 1 every time you enter an object or array.

#### Properties

`.name: "CborMaxDepthReachedError"`
:   The name of the error.

`.maxDepth: number`
:   The maximum depth.

#### Resolution

Relax the optional `.maxDepth` upper limit or revise the data structure to make the object nesting shallower.

```ts
import CborFormatter from "@tai-kun/surrealdb/cbor-formatter";

const cborFormatter = new CborFormatter({
  encode: {
    maxDepth: 1024,
  },
  decode: {
    maxDepth: 1024,
  },
  // ...
})
```

### CborUnsafeMapKeyError

Inherits from: `CborError`

This is an error thrown when an unsafe map key is found during CBOR encoding or decoding.

#### Properties

`.name: "CborUnsafeMapKeyError"`
:   The name of the error.

`.key: unknown`
:   The map key that was determined to be unsafe.

## JSON

### JsonError

Inherits from: `SurrealError`

This is a JSON-formatter explicitly thrown by `@tai-kun/surrealdb/json`.

#### Properties

`.name: "JsonError"`
:   The name of the error.

### JsonUnsafeMapKeyError

Inherits from: `JsonError`

This is an error thrown when an unsafe map key is found during JSON decoding.

#### Properties

`.name: "JsonUnsafeMapKeyError"`
:   The name of the error.

`.key: unknown`
:   The map key that was determined to be unsafe.
